### PR TITLE
Issue #2376 - relax ContextHandler requirements

### PR DESCRIPTION
**Pull request by joakime**
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java [hunk 1]
```diff
@@ -303,27 +313,76 @@ public Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse
...
-        
+    
         Objects.requireNonNull(this.objectFactory, DecoratedObjectFactory.class.getName());
         Objects.requireNonNull(this.executor, Executor.class.getName());
         
         super.doStart();
     }
-    
+
+    private DecoratedObjectFactory findDecorator()
```
Comment (gregw) [inline jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: perhaps make this protected so users could extend to add other ways to find decorators

File: jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java [hunk 2]
```diff
@@ -303,27 +313,76 @@ public Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse
...
+        objectFactory.addDecorator(new DeprecationWarning());
+        LOG.info("No DecoratedObjectFactory provided, using new {}", objectFactory);
+        return objectFactory;
+    }
+
+    /**
+     * Attempt to find the Executor that should be used.
+     * @return the Executor that should be used.
+     */
+    private Executor findExecutor()
```
Comment (gregw) [inline jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: ditto on protected

File: jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java [hunk 3]
```diff
@@ -303,27 +313,76 @@ public Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse
...
+    {
+        if(context != null)
+        {
+            // Attempt to pull Executor from ServletContext attribute
+            Executor contextExecutor = (Executor) context.getAttribute("org.eclipse.jetty.server.Executor");
```
Comment (gregw) [inline jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: Is this a new attribute?  Should be a `public final static String` and documented.
Is it really the `server.Executor`?  it's more of a  `org.eclipse.jetty.context.Executor`

Comment (joakime) [reply jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: This one already exists (just using it).
https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L766

Documented in javadoc here.
https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L104

Comment (joakime) [reply jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: Heh, a blame shows that name is something you assigned back in 2013.

I would argue that an Executor is a server thing, created/owned/managed by the server.
It's not a context feature or component, that's for sure.

Even if all we have is a context, and we want the ThreadPool the most common code is ...

``` java
Executor threadPool = context.getServer().getThreadPool();
```

Comment (gregw) [reply jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java]: Ah so it is a server threadpool, as it is the servers threadpool being exposed to the application (we should probably make that optional:)
So name is good, should have been in a public final static. can you fix?

**Pull request merged**